### PR TITLE
Remove `sharedApplication` reference from JSQMessagesBubbleImageFactory. Closes #1600

### LINKED ISSUE #1600: Remove `sharedApplication` reference from JSQMessagesBubbleImageFactory
Introduced in PR #1550.

Instead, let's:

- Add a private property for `userInterfaceLayoutDirection`
- Add this property to the designated initializer
- Use the `sharedApplication` value in the default `init`
- Update unit tests and demo

### PR DESCRIPTION
**Pull request by MacMeDan**
## Pull request checklist

- [X] All tests pass. Demo project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines), and reviewed the [contributing guidelines](https://github.com/jessesquires/JSQMessagesViewController/blob/develop/.github/CONTRIBUTING.md). Confirmation: 💪😎👊

## What's in this pull request?

> Add display images default based on `UIUserInterfaceLayoutDirection` and configurable property to overwrite 
> Tests for flipping images based on language direction.

### CODE REVIEW

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h [hunk 1]
```diff
@@ -21,12 +21,19 @@
...
+
 /**
  *  `JSQMessagesBubbleImageFactory` is a factory that provides a means for creating and styling 
  *  `JSQMessagesBubbleImage` objects to be displayed in a `JSQMessagesCollectionViewCell` of a `JSQMessagesCollectionView`.
  */
 @interface JSQMessagesBubbleImageFactory : NSObject
 
+
+/**
+ *  Property for the layout direction of text by default uses ShareedApplication.userInterfaceLayoutDirection()
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h]: Should be:

> Specifies the layout direction of the message bubble. The default value is initialized from` [UIApplication sharedApplication]`.

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h [hunk 2]
```diff
@@ -21,12 +21,19 @@
...
+ */
+@property (nonatomic)UIUserInterfaceLayoutDirection *layoutDirection;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h]: not sure why this is a pointer. should be:

```
@property (nonatomic, assign) UIUserInterfaceLayoutDirection layoutDirection;
```

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h [hunk 3]
```diff
@@ -49,7 +56,8 @@
  *
  *  @return An initialized `JSQMessagesBubbleImageFactory` object if created successfully, `nil` otherwise.
  */
-- (instancetype)initWithBubbleImage:(UIImage *)bubbleImage capInsets:(UIEdgeInsets)capInsets;
+
+- (instancetype)initWithBubbleImage:(UIImage *)bubbleImage capInsets:(UIEdgeInsets)capInsets layoutDirection:(UIUserInterfaceLayoutDirection)layoutDirection;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h]: 👍

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 4]
```diff
@@ -49,14 +49,15 @@ - (instancetype)initWithBubbleImage:(UIImage *)bubbleImage capInsets:(UIEdgeInse
         else {
             _capInsets = capInsets;
         }
+        
+        _isRightToLeftLanguage = (UIUserInterfaceLayoutDirectionRightToLeft == layoutDirection);
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: let's just save the value directly `_layoutDirection = layoutDirection;`

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 5]
```diff
@@ -49,14 +49,15 @@ - (instancetype)initWithBubbleImage:(UIImage *)bubbleImage capInsets:(UIEdgeInse
...
 	}
 	return self;
 }
 
 - (instancetype)init
 {
-  _isRightToLeftLanguage = [UIApplication sharedApplication].userInterfaceLayoutDirection == UIUserInterfaceLayoutDirectionRightToLeft;
-  return [self initWithBubbleImage:[UIImage jsq_bubbleCompactImage] capInsets:UIEdgeInsetsZero];
+  return [self initWithBubbleImage:[UIImage jsq_bubbleCompactImage] capInsets:UIEdgeInsetsZero layoutDirection:[UIApplication sharedApplication].userInterfaceLayoutDirection];
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: 👍

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m (hunk 4 above)
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: `isRightToLeftLanguage` can become a `readonly` computed property for convenience

**Comment by jessesquires** [comment]
Thanks @MacMeDan ! 🎉 

Just a few comments to address

**Comment by MacMeDan** [comment]
@jessesquires Updated the only place I had issues with was the interface 
`@interface JSQMessagesBubbleImageFactory ()

@property (strong, nonatomic, readonly) UIImage *bubbleImage;
@property (assign, nonatomic, readonly) UIEdgeInsets capInsets;
@end`

Didn't know if I was suppose to add them in here or not. My objective C is a little rusty. 👴🏼

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h [hunk 6]
```diff
@@ -30,9 +29,16 @@
...
+ * from [UIApplication sharedApplication].
+ */
+@property (nonatomic, assign)UIUserInterfaceLayoutDirection layoutDirection;
+
+/**
+ *
+ *
  */
-@property (nonatomic)UIUserInterfaceLayoutDirection *layoutDirection;
+@property (assign, readonly) BOOL isRightToLeftLanguage;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h]: let's remove this from the header

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h [hunk 7]
```diff
@@ -30,9 +29,16 @@
 
 
 /**
- *  Property for the layout direction of text by default uses ShareedApplication.userInterfaceLayoutDirection()
+ * Specifies the layout direction of the message bubble. The default value is initialized
+ * from [UIApplication sharedApplication].
+ */
+@property (nonatomic, assign)UIUserInterfaceLayoutDirection layoutDirection;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.h]: space after `assign)` plz 😁

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 8]
```diff
@@ -26,7 +26,6 @@ @interface JSQMessagesBubbleImageFactory ()
 
 @property (strong, nonatomic, readonly) UIImage *bubbleImage;
 @property (assign, nonatomic, readonly) UIEdgeInsets capInsets;
-@property (assign, nonatomic) BOOL isRightToLeftLanguage;
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: this should be:

```objc
@property (assign, nonatomic, readonly) BOOL isRightToLeftLanguage;
```

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 9]
```diff
@@ -46,11 +45,13 @@ - (instancetype)initWithBubbleImage:(UIImage *)bubbleImage capInsets:(UIEdgeInse
         if (UIEdgeInsetsEqualToEdgeInsets(capInsets, UIEdgeInsetsZero)) {
             _capInsets = [self jsq_centerPointEdgeInsetsForImageSize:bubbleImage.size];
         }
+
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: remove this newline

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 10]
```diff
@@ -64,6 +65,7 @@ - (instancetype)init
 
 - (JSQMessagesBubbleImage *)outgoingMessagesBubbleImageWithColor:(UIColor *)color
 {
+
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: remove this newline

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m [hunk 11]
```diff
@@ -72,6 +74,14 @@ - (JSQMessagesBubbleImage *)incomingMessagesBubbleImageWithColor:(UIColor *)colo
     return [self jsq_messagesBubbleImageWithColor:color flippedForIncoming:YES ^ self.isRightToLeftLanguage];
 }
 
+- (BOOL)isRightToLeftLanguage:(UIUserInterfaceLayoutDirection)layoutDirection {
+    if (layoutDirection == UIUserInterfaceLayoutDirectionRightToLeft) {
+        return true;
+    } else {
+    return false;
+    }
+}
```
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: this should be:

```objc
- (BOOL)isRightToLeftLanguage {
    return (self.layoutDirection == UIUserInterfaceLayoutDirectionRightToLeft);
}
```

File: JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m [hunk 12]
```diff
@@ -136,13 +136,12 @@ - (void)testCustomOutgoingMessageBubbleImageViewWithLeftToRigthtDirectionText
     XCTAssertNotNil(bubbleImage, @"Bubble image should not be nil");
     XCTAssertNotNil(bubbleImage.messageBubbleImage, "Image should not be nil");
     XCTAssertEqual(bubbleImage.messageBubbleImage.scale, bubble.scale, @"Image scale should equal bubble image scale");
-    XCTAssertNotEqual(bubbleImage.messageBubbleImage.imageOrientation, bubble.imageOrientation, @"Image orientation should not equal bubble image orientation");
+    XCTAssertEqual(bubbleImage.messageBubbleImage.imageOrientation, bubble.imageOrientation, @"Image orientation should not equal bubble image orientation");
```
Comment (jessesquires) [inline JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m]: this shouldn't have needed to change

File: JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m [hunk 13]
```diff
@@ -136,13 +136,12 @@ - (void)testCustomOutgoingMessageBubbleImageViewWithLeftToRigthtDirectionText
...
     XCTAssertTrue(bubbleImage.messageBubbleImage.resizingMode == UIImageResizingModeStretch, @"Image should be stretchable");
     XCTAssertTrue(UIEdgeInsetsEqualToEdgeInsets(bubbleImage.messageBubbleImage.capInsets, capInsets), @"Image capInsets should be equal to capInsets");
     XCTAssertFalse(bubbleImage.messageBubbleImage.flipsForRightToLeftLayoutDirection, @"Not the right direction");
     XCTAssertNotNil(bubbleImage.messageBubbleHighlightedImage, @"Highlighted image should not be nil");
     XCTAssertEqual(bubbleImage.messageBubbleHighlightedImage.scale, bubble.scale, @"HighlightedImage scale should equal bubble image scale");
-    XCTAssertNotEqual(bubbleImage.messageBubbleHighlightedImage.imageOrientation, bubble.imageOrientation, @"HighlightedImage orientation should be the same bubble image orientation");
```
Comment (jessesquires) [inline JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m]: this shouldn't have needed to change

File: JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m [hunk 14]
```diff
@@ -180,7 +179,7 @@ - (void)testCustomIncomingMessageBubbleImageViewWithLeftToRigthtDirectionText
     
     XCTAssertNotNil(bubbleImage.messageBubbleImage, "Image should not be nil");
     XCTAssertEqual(bubbleImage.messageBubbleImage.scale, bubble.scale, @"Image scale should equal bubble image scale");
-    XCTAssertNotEqual(bubbleImage.messageBubbleImage.imageOrientation, UIImageOrientationUpMirrored, @"Image orientation should be flipped");
+    XCTAssertEqual(bubbleImage.messageBubbleImage.imageOrientation, UIImageOrientationUpMirrored, @"Image orientation should be flipped");
```
Comment (jessesquires) [inline JSQMessagesTests/FactoryTests/JSQMessagesBubbleImageFactoryTests.m]: this shouldn't have needed to change

File: JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m (hunk 11 above)
Comment (jessesquires) [inline JSQMessagesViewController/Factories/JSQMessagesBubbleImageFactory.m]: what you wrote here was an entirely new method, not an override for the property.

thus `self.isRightToLeftLanguage` was always `NO`. (which is probably why you had to change the tests) 😅

**Comment by jessesquires** [comment]
Thanks @MacMeDan ! A few more comments 😄

**Comment by MacMeDan** [comment]
Ok @jessesquires I think I have it all straightened out.

**Comment by jessesquires** [comment]
Thanks @MacMeDan !

A few more minor things, but I'll take care of it 👍

**Comment by owenmills** [comment]
is it just me, or is there still a reference to sharedApplication in JSQMessagesBubbleImageFactory.m on line 58?
`return [self initWithBubbleImage:[UIImage jsq_bubbleCompactImage] capInsets:UIEdgeInsetsZero layoutDirection:[UIApplication sharedApplication].userInterfaceLayoutDirection];`

**Comment by jessesquires** [comment]
@owenmills - yes, the goal here was to:

1. provide 1 full designated initializer
2. provide a sensible default `init` to avoid pushing too much boilerplate on clients

**Comment by owenmills** [comment]
@jessesquires oh right, my apologies.
I stumbled across this because I was trying to integrate this framework into a messages extension app which will be available in xcode 8 / ios10.  One of the restrictions is that messages extensions don't have access to sharedApplication.  I wrongly assumed this ticket was trying to remove all references to enable support for messages extension.

**Comment by jessesquires** [comment]
@owenmills Ah, right. Hm... supporting both of these scenarios would require quite a bit of work right now

**Pull request merged**